src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1248,6 +1220,123 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
             }
         }
 
+        private bool IsRefStruct(out bool isRefFollowedByStructOrPartialStruct, out bool isTypeDeclaration)
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.RefKeyword);
+
+            var nextToken = PeekToken(1);
+
+            isRefFollowedByStructOrPartialStruct = nextToken.Kind == SyntaxKind.StructKeyword ||
+                (nextToken.ContextualKind == SyntaxKind.PartialKeyword && this.PeekToken(2).Kind == SyntaxKind.StructKeyword);
+
+            if (isRefFollowedByStructOrPartialStruct)
+            {
+                isTypeDeclaration = true;
+                return true;
+            }
+
+            isTypeDeclaration = IsTypeDeclaration(out nextToken);
+            return isTypeDeclaration && nextToken.Kind == SyntaxKind.StructKeyword;
+        }
+
+        private static bool CanFollowTypeName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.DotToken:                   // partial.     qualified
+                case SyntaxKind.ColonColonToken:            // partial::    alias
+                case SyntaxKind.LessThanToken:              // partial<     generic type
+                case SyntaxKind.OpenBracketToken:           // partial[     array
```

Comment:
the containing function is very vaguely named given that the comments seem to imply this is all about hwat's allowed after 'partial' if 'partial' were a type name.  I would have never expected that given the name.

Reply:
could work for anything else but we used it only on contextual keywords.